### PR TITLE
[KYUUBI #1446] Decouple zookeeper from abstract ServiceDiscovery

### PR DESCRIPTION
**Pull request by hddong**
### _Why are the changes needed?_

Decouple zookeeper from abstract ServiceDiscovery

### _How was this patch tested?_
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
can we revert the unrelated change first?

**Comment by hddong** [comment]
> can we revert the unrelated change first?

Reverted

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZooKeeperInstanceClient.scala [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  def beforeStop(): Unit = {
+    closeServiceNode()
+  }
+
+  def deletingChildrenIfNeeded(path: String): Unit = {
+    zkClient.delete().deletingChildrenIfNeeded().forPath(namespace)
+  }
+
+  // close the EPHEMERAL_SEQUENTIAL node in zk
+  private def closeServiceNode(): Unit = {
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZooKeeperInstanceClient.scala:118]: how about we expose closeServiceNode directly?

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZooKeeperInstanceClient.scala [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+import org.apache.kyuubi.{KyuubiException, Logging}
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.ha.HighAvailabilityConf.HA_ZK_CONN_MAX_RETRIES
+import org.apache.kyuubi.ha.HighAvailabilityConf.HA_ZK_CONN_MAX_RETRY_WAIT
+import org.apache.kyuubi.ha.HighAvailabilityConf.HA_ZK_NAMESPACE
+import org.apache.kyuubi.ha.client.ServiceDiscovery.createServiceNode
+import org.apache.kyuubi.ha.client.ZooKeeperClientProvider.buildZookeeperClient
+import org.apache.kyuubi.util.ThreadUtils
+
+class ZooKeeperInstanceClient extends Logging {
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZooKeeperInstanceClient.scala:43]: We have a zookeeper in package name already, and we use the client to interact with serviceDiscovery. Som How about 
```scala
class zookeeper.ServiceDiscoveryClient(serviceDiscovery: ServiceDiscovery) {
  def createClient.. // initialize
  def closeClient.. // stop
  // def preRegisterService
  def registerService.. // start
  // def postRegisterService
  // def preDeregisterService
  def deregisterService... // beforeStop, closeServiceNode
  def postDeregisterService ... // deletingChildrenIfNeeded
}
```

Comment (yaooqinn) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZooKeeperInstanceClient.scala:43]: or class ZookeeperServiceDiscoveryClient and remove zookeeper in package

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZooKeeperInstanceClient.scala [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  protected var _namespace: String = _
+
+  private[client] var instance: String = _
+  private var zkClient: CuratorFramework = _
+  private var serviceNode: PersistentNode = _
+
+  def namespace: String = _namespace
+  protected var serviceStopGracefully: () => Unit = _
+
+  final private lazy val connectionChecker =
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZooKeeperInstanceClient.scala:57]: Before, this is a singleton

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZooKeeperInstanceClient.scala (hunk 2 above)
Comment (yaooqinn) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZooKeeperInstanceClient.scala:43]: a typically calling order shall be
 `new ServiceDiscoveryClient` ->
 `createClient` ->
 `preRegisterService(if any)` -> `registerService` -> `postRegisterService (if any)`  ->
 `preDeregisterService` -> `deregisterService ` -> `postDeregisterService ` -> 
`closeClient`

**Comment by hddong** [comment]
@yaooqinn : Thanks for your review, and had address them.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala [hunk 4]
```diff
@@ -48,92 +48,34 @@ import org.apache.kyuubi.util.{KyuubiHadoopUtils, ThreadUtils}
...
-  import ZooKeeperClientProvider._
-
-  private var _zkClient: CuratorFramework = _
-  private var _serviceNode: PersistentNode = _
-
-  /**
-   * a pre-defined namespace used to publish the instance of the associate service
-   */
+  private var _discoveryClient: ServiceDiscoveryClient = _
   private var _namespace: String = _
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:54]: we don't need this anymore?

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala [hunk 5]
```diff
@@ -48,92 +48,34 @@ import org.apache.kyuubi.util.{KyuubiHadoopUtils, ThreadUtils}
...
 
-  def zkClient: CuratorFramework = _zkClient
-
-  def serviceNode: PersistentNode = _serviceNode
-
+  def discoveryClient: ServiceDiscoveryClient = _discoveryClient
   def namespace: String = _namespace
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:57]: ditto

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/ServiceDiscoverySuite.scala [hunk 6]
```diff
@@ -29,7 +29,8 @@ import org.apache.zookeeper.ZooDefs
 import org.apache.zookeeper.data.ACL
 import org.scalatest.time.SpanSugar._
 
-import org.apache.kyuubi.{KerberizedTestHelper, KYUUBI_VERSION}
+import org.apache.kyuubi.KerberizedTestHelper
+import org.apache.kyuubi.KYUUBI_VERSION
```
Comment (yaooqinn) [inline kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/ServiceDiscoverySuite.scala:33]: unnecessary

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ServiceDiscoveryClient.scala [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    // Set a watch on the serviceNode
+    val watcher = new DeRegisterWatcher
+    if (zkClient.checkExists.usingWatcher(watcher).forPath(serviceNode.getActualPath) == null) {
+      // No node exists, throw exception
+      throw new KyuubiException(s"Unable to create znode for this Kyuubi " +
+        s"instance[${instance}] on ZooKeeper.")
+    }
+  }
+
+  // close the EPHEMERAL_SEQUENTIAL node in zk
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ServiceDiscoveryClient.scala:102]: let make some comment more common for this func, and move the current comment inside the func

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ServiceDiscoveryClient.scala [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+              TimeUnit.MILLISECONDS)
+          case _ =>
+        }
+      }
+    })
+    zkClient.start()
+  }
+
+  def registerService(conf: KyuubiConf): Unit = {
+    serviceNode = createServiceNode(conf, zkClient, namespace, instance)
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ServiceDiscoveryClient.scala:92]: shall we move the entire createServiceNode here?

**Comment by yaooqinn** [comment]
LGTM overall, only minors

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**